### ISSUE TITLE
v4.0.x compile failure under pre C99 standard

### ISSUE THREAD

**Issue by AdamSimpson** [open-mpi/ompi#5893]
### What version of Open MPI are you using? (e.g., v1.10.3, v2.1.0, git branch name and hash, etc.)
v4.0.x branch

## Details of the problem
`mpi.h` contains C99/C++ style comments, `//`, when MPI is compiled with `--enable-mpi1-compatibility`. This breaks compilation under pre C99 standards.

Offending section of `mpi.h`
```
// NOTE: We intentionally do *not* mark the following as
// deprecated/removed because they are used below in function
// prototypes (and would therefore emit warnings, just for #including
// <mpi.h>).
```
Example failure
```
$ cat hello.c
#include "mpi.h"

int main(int argc, char** argv) {
    return 0;
}

$ mpicc -ansi hello.c
/usr/local/openmpi/include/mpi.h:420:1: error: C++ style comments are not allowed in ISO C90
 // NOTE: We intentionally do *not* mark the following as
 ^
/usr/local/openmpi/include/mpi.h:420:1: error: (this will be reported only once per input file)
```

**Comment by ggouaillardet** [comment]
`C99` is a requirement in order to __build__ Open MPI.

per your report, the `mpicc` wrapper is failing, and I am not sure whether this is a bug or a feature.

@hjelmn @jsquyres could you please shed some light ?

**Comment by hjelmn** [comment]
We shouldn't have c++ style comments in mpi.h

**Comment by jsquyres** [comment]
I had the same knee-jerk reaction: we require C99 to build Open MPI.

...but that's to build Open MPI.  We do not require C99 for user application.  @hjelmn is right -- we shouldn't have C99 constructs (e.g., `\\`-style comments) in `mpi.h[.in]`.

I hate to say it, but that looks like a Jeff-style comment (i.e., that wording looks like something I would say in a comment) -- I'm a little afraid to run `git blame` here 😳

I'll just go prepare a PR right now to fix the issue...

**Comment by jsquyres** [comment]
Confirmed that this affects master and v4.0.x, but not earlier releases.

**Comment by jsquyres** [comment]
This has been merged to the v4.0.x release branch for the upcoming v4.0.0 release.
